protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java

```diff
@@ -0,0 +1,305 @@
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.remove(messageUid1);
+
+        assertThat(testee.getMsn(messageUid1).isPresent()).isFalse();
+        assertThat(testee.getMsn(messageUid2).isPresent()).isTrue();
+        assertThat(testee.getMsn(messageUid3).isPresent()).isTrue();
+    }
+
+    @Test
+    public void  removeShouldRemoveOnlyTheCorrectUidWhenOneTheEnd() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.remove(messageUid3);
+
+        assertThat(testee.getMsn(messageUid3).isPresent()).isFalse();
+        assertThat(testee.getMsn(messageUid1).isPresent()).isTrue();
+        assertThat(testee.getMsn(messageUid2).isPresent()).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepMessageNumberContiguousWhenDeletingTheFirstOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid1);
+
+        assertThat(messageNumberAreContiguous(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepMessageNumberContiguousWhenDeletingTheLastOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid4);
+
+        assertThat(messageNumberAreContiguous(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepMessageNumberContiguousWhenDeletingOneInTheMiddle() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid3);
+
+        assertThat(messageNumberAreContiguous(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepUidInOrderWhenDeletingTheFirstOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid1);
+
+        assertThat(uidAreInOrder(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepUidInOrderWhenDeletingTheLastOne() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid4);
+
+        assertThat(uidAreInOrder(testee)).isTrue();
+    }
+
+    @Test
+    public void removeShouldKeepUidInOrderWhenDeletingOneInTheMiddle() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+
+        testee.remove(messageUid3);
+
+        assertThat(uidAreInOrder(testee)).isTrue();
+    }
+
+    private static boolean messageNumberAreContiguous(UidMsnMapper testee) {
+        int first = UidMsnMapper.FIRST_MSN;
```

Comment:
2. No. Sorry. You should keep tests as simple as possible. No concept extraction : you should check data in your unit test. This is what is important. Btw exposing your mapping as a map is IMO not implem depandant...
If you check it with 4 asserts everywhere, or with an assert on a @VisibleForTesting collection (which is common pattern in james team), then be it. I should not need this complex contiguous thing to understand what your class is doing.